src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.Unix.cs

```diff
@@ -321,7 +316,24 @@ private unsafe void SetAccessOrWriteTime(string path, DateTimeOffset time, bool
             }
 #endif
             Interop.CheckIo(Interop.Sys.UTimensat(path, buf), path, InitiallyDirectory);
-            _initializedFileCache = -1;
+
+            // On OSX platforms, when the modification time is less than the creation time, the creation time
+            // is broken; so these api calls revert the creation time when it shouldn't be set (since we're
+            // setting modification time and access time here). checkCreationTime is only true on OSX
+            // platforms. allowFallbackToLastWriteTime is ignored on non OSX platforms.
```

Comment:
> This code will be called not only for LastWriteTime (modification time), but also for LastAccessTime, do we want to update the creation time in that case as well?

Modification and access time are set together, as far as the OS is concerned, they are both set.

> What do you mean by broken?

It changes the creation time as well, I'll update this comment to be more clear about what it does.

> Also, how did you find out about this issue?
Is there any doc describing that you should update creation time if modification time is set to an older date i.e: is this documented by apple?
Did you add unit tests related to the described scenario?

I found out about this scenario when testing my copy of the code locally in a project I develop (I have most of this PR in that project at the moment). There are tests for this (it's the complicated test with 12 different options).
I think I saw it documented somewhere, but I cannot remember where, I'm pretty sure it's because we're using the standard unix api instead of the standard bsd api (which we're now using for creation time). (From memory:) If we use the standard unix api it does modification and creation correctly but does something odd with access time, so it's a non-solution (unless you can work around it I guess). Perhaps we can investigate that in the future, but I don't really see any point.

Just to be really clear (for possible future reference): when the api that is called by `Interop.Sys.UTimensat` is called, the modification and access times are both set, and if the modification date is less than the creation time: then the creation time is also set to the modification time.

Reply:
> > This code will be called not only for LastWriteTime (modification time), but also for LastAccessTime, do we want to update the creation time in that case as well?
> 
> Modification and access time are set together, as far as the OS is concerned, they are both set.

Yes, but changing access time doesn't change creation time automatically.
e.g:
```
info.LastWriteTime = DateTime.Now;
info.CreationTime = DateTime.Now.AddDays(-1); // yesterday
info.LastAccessTime = DateTime.Now.AddDays(-2); // the day before yesterday

Console.WriteLine($"Last Modified: {info.LastWriteTime"); // today
Console.WriteLine($"Created: {info.CreationTime}"); // today - unexpected
Console.WriteLine($"Last Access: {info.LastAccessTime"); // the day before yesterday
```
with your proposed change it will change it and that would be a breaking change.

Reply:
Hi, your example is incorrect. This behaviour will never occur with or without my change.
All that my code fixes is the following:

Without it:
```cs
info.LastWriteTime = DateTime.Now.AddDays(-2); // the day before yesterday
info.CreationTime = DateTime.Now.AddDays(-1); // yesterday
info.LastAccessTime = DateTime.Now; // today

Console.WriteLine($"Last Modified: {info.LastWriteTime"); // the day before yesterday
Console.WriteLine($"Created: {info.CreationTime}"); // the day before yesterday - unexpected
Console.WriteLine($"Last Access: {info.LastAccessTime"); // today
```

With it:
```cs
info.LastWriteTime = DateTime.Now.AddDays(-2); // the day before yesterday
info.CreationTime = DateTime.Now.AddDays(-1); // yesterday
info.LastAccessTime = DateTime.Now; // today

Console.WriteLine($"Last Modified: {info.LastWriteTime"); // the day before yesterday
Console.WriteLine($"Created: {info.CreationTime}"); // yesterday
Console.WriteLine($"Last Access: {info.LastAccessTime"); // today
```

There are multiple reasons to desire this (2) behaviour, mainly windows compatibility and not surprising the developer that they need to always set creation time last on OSX-like platforms (for which many wouldn't have a device to test on).
Does that make sense?

Reply:
Also, it can't be a breaking change since you couldn't set creation time in .NET on OSX-like platforms before.

Reply:
It appers that `UTimensat` changes the creation time as well for this scenario, although, I am not sure how valid it is to use `default` as a date in Unix. 

```cs
var info = new FileInfo("foo");

// Reset all timestamps:
info.LastWriteTime = default;
info.CreationTime = default;
info.LastAccessTime = default;

// Verify all of them are in the same date ~9/20/1677 4:19:43 PM
Console.WriteLine($"Last Modified: {info.LastWriteTime}"); 
Console.WriteLine($"Created: {info.CreationTime}"); 
Console.WriteLine($"Last Access: {info.LastAccessTime}"); 

// Change one date.
info.LastWriteTime = DateTime.Now;

// Verify that changing one date doesn't affect the other dates
Console.WriteLine($"Last Modified: {info.LastWriteTime}"); 
Console.WriteLine($"Created: {info.CreationTime}"); // It changed when it wasn't expected to change.
Console.WriteLine($"Last Access: {info.LastAccessTime}"); 
```

What do you think? Should we do extensive cover and find all scenarios where `Utimensat` changes creation time and revert it. Or does the current condition suffice?

Another option could be to not try to revert the creation date and keep the current behavior of "UTimensat may change your creation date". 
It would be good to find out how this case is handled somewhere else.

Reply:
It appers that `UTimensat` changes the creation time as well for this scenario, although, I am not sure how valid it is to use `default` as a timestamp in Unix. 

```cs
var info = new FileInfo("foo");

// Reset all timestamps:
info.LastWriteTime = default;
info.CreationTime = default;
info.LastAccessTime = default;

// Verify all of them are in the same date ~9/20/1677 4:19:43 PM
Console.WriteLine($"Last Modified: {info.LastWriteTime}"); 
Console.WriteLine($"Created: {info.CreationTime}"); 
Console.WriteLine($"Last Access: {info.LastAccessTime}"); 

// Change one date.
info.LastWriteTime = DateTime.Now;

// Verify that changing one date doesn't affect the other dates
Console.WriteLine($"Last Modified: {info.LastWriteTime}"); 
Console.WriteLine($"Created: {info.CreationTime}"); // It changed when it wasn't expected to change.
Console.WriteLine($"Last Access: {info.LastAccessTime}"); 
```

What do you think? Should we do extensive cover and find all scenarios where `Utimensat` changes creation time and revert it. Or does the current condition suffice?

Another option could be to not try to revert the creation date and keep the current behavior of "UTimensat may change your creation date". 
It would be good to find out how this case is handled somewhere else.

Reply:
> It appers that `UTimensat` changes the creation time as well for this scenario, although, I am not sure how valid it is to use `default` as a timestamp in Unix.

I wouldn't be surprised if it isn't a "valid" concept since unix timestamps acceptable range can vary based on the os (eg. on apple it is `_MAX__TIME64_T = 0x793406fffi64`).

Though I'm not exactly sure what `default` means for a timestamp personally.

> What do you think? Should we do extensive cover and find all scenarios where `Utimensat` changes creation time and revert it. Or does the current condition suffice?

I think we should keep the check in the code and revert it. I think the current condition should cover everything, if it doesn't, we can always update it in the future, but I think it's unlikely.

Also, with your test case, I think it would put the creation time back since those seem like reasonable values for unix and the comparison operator should be able to work it out - I can add a test if you'd like.

Reply:
Yes, add the tests for it, please.

Reply:
I will put this in the next commit.